### PR TITLE
Extend 'BlockHeader' type with current 'blockHash'

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#219 

# Overview

- [ ] I have extended 'BlockHeader' with the current 'blockHash'. This will make it more feasible to implement 'getDescendentIds' with Jörmungandr.
- [ ] I have removed the hard-coded genesis block from the wallet model and have it passed as argument. This depends on the underlying network actually ... 
- [ ] I have fixed remaining compile-error and some styling issues with the base PR.

# Comments

:warning: Target branch is `anviking/219/blockheader-tip` :warning:

### CODE REVIEW

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Binary.hs [hunk 1]
```diff
@@ -203,7 +203,10 @@ decodeGenesisBlockHeader = do
     -- number of `0`. In practices, when parsing a full epoch, we can discard
     -- the genesis block entirely and we won't bother about modelling this
     -- extra complexity at the type-level. That's a bit dodgy though.
-    return $ BlockHeader (SlotId epoch 0) previous
+    return $ BlockHeader
+        (SlotId epoch 0)
+        (Hash "http-bridge blockHash not implemented")
```
Comment (Anviking) [inline lib/http-bridge/src/Cardano/Wallet/HttpBridge/Binary.hs:208]: Are we okay with filling the `Sqlite` database with this string?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 2]
```diff
@@ -397,7 +418,10 @@ instance FromBinary W.Block where
             W.Block (convertHeader h) (convertMessages msgs)
 
         convertHeader :: BlockHeader -> W.BlockHeader
-        convertHeader h = W.BlockHeader (slot h) (parentHeaderHash h)
+        convertHeader h = W.BlockHeader
+            (slot h)
+            (blake2b256 $ putBlockHeader h)
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:423]: Is it not a performance penalty to re-encode and hash every block-header we decode?

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs [hunk 3]
```diff
@@ -124,3 +127,18 @@ instance DecodeAddress (HttpBridge (network :: Network)) where
...
         errDecoding _ = "Unable to decode Address: not a valid Byron address."
+
+-- | An initial first block to initialize a chain using the http-bridge. We do
+-- not use the `blockHash` and, do only use the `prevBlockHash` to catch up with
+-- unstable epoch and therefore, the very first `prevBlockHash` matters not.
+--
+-- It isn't impossible to retrieve the 'blockHash' by computing a blake2b 256 of
+-- the CBOR-serialized full block header, but this requires us to write the full
+-- CBOR decoders (and encoders) for the all BlockHeader which is, for the
+-- http-brdige implementation, a waste of time at the moment.
```
Comment (Anviking) [inline lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs:138]: In that case — looking at a chunk of blocks and getting the `prevBlockHash` of the child/next block might be an idea.

File: lib/core/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -368,7 +370,7 @@ newWalletLayer db nw tl = do
         -> s
         -> ExceptT ErrWalletAlreadyExists IO WalletId
     _createWallet wid wname s = do
-        let checkpoint = initWallet s
+        let checkpoint = initWallet block0 s
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:373]: I take it we don't have to `applyBlock` `block0`?

**Review by Anviking — COMMENTED** [review]
> This will make it more feasible to implement 'getDescendentIds' with Jörmungandr

I know I just believed that myself too, but what is your specific reason?

Most recently I was thinking we could manage with just a `BlockHeader` by implementing the `Jörmungandr` `NetworkLayer` `nextBlocks` like this:
```haskell
nextBlocks = \h@(BlockHeader slot prevBlockHash) -> do
    if  h == (block0 ^. #blockHeader)
    then -- prepend block0 to the results
    else do
        -- get the descendants of the tip's parent
        -- the first descendant is therefore the tip itself
        (_curBlockId:ids) <- getDescendantIds prevBlockHash 1000
        forM ids getBlock'
```

and thus not have to re-encode and hash the `BlockHeader`.

**Comment by Anviking** [comment]
Thanks for the awesome commit history btw 🙏

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Binary.hs (hunk 1 above)
Comment (KtorZ) [reply lib/http-bridge/src/Cardano/Wallet/HttpBridge/Binary.hs:208]: Yes. We don't have that many block headers / checkpoints in DB (even with rollbacks)

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs (hunk 2 above)
Comment (KtorZ) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:423]: A bit, perhaps. Can't tell without measuring. It's a hash relatively fast to compute, but that's an overhead for sure.

**Comment by KtorZ** [comment]
> Most recently I was thinking we could manage with just a BlockHeader by implementing the Jörmungandr NetworkLayer nextBlocks like this:

That's a very good point. So maybe, not everything in this PR is needed. If we can do without the current block header hash then that's better.

**Comment by KtorZ** [comment]
Relevant commits cherry-picked on the base branch, closing this PR.

**Pull request closed without merging**